### PR TITLE
Arm64: Implement LoadVector64x* and LoadVector128x* APIs

### PR DESCRIPTION
**Pull request by kunalspathak**
Add 51 `LoadVector*` APIs

```c#
public static unsafe (Vector128<byte> Value1, Vector128<byte> Value2) LoadVector128x2(byte* address);
public static unsafe (Vector128<sbyte> Value1, Vector128<sbyte> Value2) LoadVector128x2(sbyte* address);
public static unsafe (Vector128<short> Value1, Vector128<short> Value2) LoadVector128x2(short* address);
public static unsafe (Vector128<ushort> Value1, Vector128<ushort> Value2) LoadVector128x2(ushort* address);
public static unsafe (Vector128<int> Value1, Vector128<int> Value2) LoadVector128x2(int* address);
public static unsafe (Vector128<uint> Value1, Vector128<uint> Value2) LoadVector128x2(uint* address);
public static unsafe (Vector128<long> Value1, Vector128<long> Value2) LoadVector128x2(long* address);
public static unsafe (Vector128<ulong> Value1, Vector128<ulong> Value2) LoadVector128x2(ulong* address);
public static unsafe (Vector128<float> Value1, Vector128<float> Value2) LoadVector128x2(float* address);
public static unsafe (Vector128<double> Value1, Vector128<double> Value2) LoadVector128x2(double* address);
public static unsafe (Vector128<byte> Value1, Vector128<byte> Value2, Vector128<byte> Value3) LoadVector128x3(byte* address);
public static unsafe (Vector128<sbyte> Value1, Vector128<sbyte> Value2, Vector128<sbyte> Value3) LoadVector128x3(sbyte* address);
public static unsafe (Vector128<short> Value1, Vector128<short> Value2, Vector128<short> Value3) LoadVector128x3(short* address);
public static unsafe (Vector128<ushort> Value1, Vector128<ushort> Value2, Vector128<ushort> Value3) LoadVector128x3(ushort* address);
public static unsafe (Vector128<int> Value1, Vector128<int> Value2, Vector128<int> Value3) LoadVector128x3(int* address);
public static unsafe (Vector128<uint> Value1, Vector128<uint> Value2, Vector128<uint> Value3) LoadVector128x3(uint* address);
public static unsafe (Vector128<long> Value1, Vector128<long> Value2, Vector128<long> Value3) LoadVector128x3(long* address);
public static unsafe (Vector128<ulong> Value1, Vector128<ulong> Value2, Vector128<ulong> Value3) LoadVector128x3(ulong* address);
public static unsafe (Vector128<float> Value1, Vector128<float> Value2, Vector128<float> Value3) LoadVector128x3(float* address);
public static unsafe (Vector128<double> Value1, Vector128<double> Value2, Vector128<double> Value3) LoadVector128x3(double* address);
public static unsafe (Vector128<byte> Value1, Vector128<byte> Value2, Vector128<byte> Value3, Vector128<byte> Value4) LoadVector128x4(byte* address);
public static unsafe (Vector128<sbyte> Value1, Vector128<sbyte> Value2, Vector128<sbyte> Value3, Vector128<sbyte> Value4) LoadVector128x4(sbyte* address);
public static unsafe (Vector128<short> Value1, Vector128<short> Value2, Vector128<short> Value3, Vector128<short> Value4) LoadVector128x4(short* address);
public static unsafe (Vector128<ushort> Value1, Vector128<ushort> Value2, Vector128<ushort> Value3, Vector128<ushort> Value4) LoadVector128x4(ushort* address);
public static unsafe (Vector128<int> Value1, Vector128<int> Value2, Vector128<int> Value3, Vector128<int> Value4) LoadVector128x4(int* address);
public static unsafe (Vector128<uint> Value1, Vector128<uint> Value2, Vector128<uint> Value3, Vector128<uint> Value4) LoadVector128x4(uint* address);
public static unsafe (Vector128<long> Value1, Vector128<long> Value2, Vector128<long> Value3, Vector128<long> Value4) LoadVector128x4(long* address);
public static unsafe (Vector128<ulong> Value1, Vector128<ulong> Value2, Vector128<ulong> Value3, Vector128<ulong> Value4) LoadVector128x4(ulong* address);        
public static unsafe (Vector128<float> Value1, Vector128<float> Value2, Vector128<float> Value3, Vector128<float> Value4) LoadVector128x4(float* address);
public static unsafe (Vector128<double> Value1, Vector128<double> Value2, Vector128<double> Value3, Vector128<double> Value4) LoadVector128x4(double* address);

public static unsafe (Vector64<byte> Value1, Vector64<byte> Value2) LoadVector64x2(byte* address);
public static unsafe (Vector64<sbyte> Value1, Vector64<sbyte> Value2) LoadVector64x2(sbyte* address);
public static unsafe (Vector64<short> Value1, Vector64<short> Value2) LoadVector64x2(short* address);
public static unsafe (Vector64<ushort> Value1, Vector64<ushort> Value2) LoadVector64x2(ushort* address);
public static unsafe (Vector64<int> Value1, Vector64<int> Value2) LoadVector64x2(int* address);
public static unsafe (Vector64<uint> Value1, Vector64<uint> Value2) LoadVector64x2(uint* address);
public static unsafe (Vector64<float> Value1, Vector64<float> Value2) LoadVector64x2(float* address);
public static unsafe (Vector64<byte> Value1, Vector64<byte> Value2, Vector64<byte> Value3) LoadVector64x3(byte* address);
public static unsafe (Vector64<sbyte> Value1, Vector64<sbyte> Value2, Vector64<sbyte> Value3) LoadVector64x3(sbyte* address);
public static unsafe (Vector64<short> Value1, Vector64<short> Value2, Vector64<short> Value3) LoadVector64x3(short* address);
public static unsafe (Vector64<ushort> Value1, Vector64<ushort> Value2, Vector64<ushort> Value3) LoadVector64x3(ushort* address);
public static unsafe (Vector64<int> Value1, Vector64<int> Value2, Vector64<int> Value3) LoadVector64x3(int* address);
public static unsafe (Vector64<uint> Value1, Vector64<uint> Value2, Vector64<uint> Value3) LoadVector64x3(uint* address);
public static unsafe (Vector64<float> Value1, Vector64<float> Value2, Vector64<float> Value3) LoadVector64x3(float* address);
public static unsafe (Vector64<byte> Value1, Vector64<byte> Value2, Vector64<byte> Value3, Vector64<byte> Value4) LoadVector64x4(byte* address);
public static unsafe (Vector64<sbyte> Value1, Vector64<sbyte> Value2, Vector64<sbyte> Value3, Vector64<sbyte> Value4) LoadVector64x4(sbyte* address);
public static unsafe (Vector64<short> Value1, Vector64<short> Value2, Vector64<short> Value3, Vector64<short> Value4) LoadVector64x4(short* address);
public static unsafe (Vector64<ushort> Value1, Vector64<ushort> Value2, Vector64<ushort> Value3, Vector64<ushort> Value4) LoadVector64x4(ushort* address);
public static unsafe (Vector64<int> Value1, Vector64<int> Value2, Vector64<int> Value3, Vector64<int> Value4) LoadVector64x4(int* address);
public static unsafe (Vector64<uint> Value1, Vector64<uint> Value2, Vector64<uint> Value3, Vector64<uint> Value4) LoadVector64x4(uint* address);
public static unsafe (Vector64<float> Value1, Vector64<float> Value2, Vector64<float> Value3, Vector64<float> Value4) LoadVector64x4(float* address);        
```

Contributes to https://github.com/dotnet/runtime/issues/84510

### CODE REVIEW

**Comment by kunalspathak** [comment]
cc: @a74nh @SwapnilGaikwad

File: src/coreclr/jit/gentree.h [hunk 1]
```diff
@@ -6085,15 +6089,66 @@ struct GenTreeJitIntrinsic : public GenTreeMultiOp
...
+    //-----------------------------------------------------------
+    // GetRegNumByIdx: Get regNumber of i'th position.
+    //
+    // Arguments:
+    //    idx   -   register position.
+    //
+    // Return Value:
+    //    Returns regNumber assigned to i'th position.
+    //
+    regNumber GetRegNumByIdx(unsigned idx) const
```
Comment (a74nh) [inline src/coreclr/jit/gentree.h:6101]: Should this (and the Set function) be inline? Or do we rely of the compiler to make that choice?

Comment (kunalspathak) [reply src/coreclr/jit/gentree.h:6101]: For other definitions of sibling classes, for this methog, we have usually relied on compiler to make that choice. So will keep it as is.

**Comment by kunalspathak** [comment]
@vargaz - any idea why `MinNumberPairwise` is failing in mono ?

```
	  at <unknown> <0xffffffff>
	  at System.Object:__icall_wrapper_mono_gc_alloc_string <0x00007>
	  at System.Object:AllocString <0x00107>
	  at System.String:Concat <0x00103>
	  at TestLibrary.Generator:GetSingle <0x0006f>
	  at JIT.HardwareIntrinsics.Arm._AdvSimd.Arm64.SimpleBinaryOpTest__MinNumberPairwise_Vector64_Single:.ctor <0x002ff>
	  at JIT.HardwareIntrinsics.Arm._AdvSimd.Arm64.Program:MinNumberPairwise_Vector64_Single <0x0003f>
	  at Program:<<Main>$>g__TestExecutor360|0_361 <0x000f3>
	  at Program:<Main>$ <0x01973>
	  at <Module>:runtime_invoke_int_object <0x00087>
```

**Comment by kunalspathak** [comment]
@dotnet/jit-contrib @BruceForstall

File: src/coreclr/jit/lsraarm64.cpp [hunk 2]
```diff
@@ -1816,6 +1829,57 @@ int LinearScan::BuildConsecutiveRegistersForUse(GenTree* treeNode, GenTree* rmwN
     return srcCount;
 }
 
+//------------------------------------------------------------------------
+//  BuildConsecutiveRegistersForDef: Build RefTypeDef ref position(s) for
+//  `treeNode` that produces `fieldCount` consecutive registers.
```
Comment (BruceForstall) [inline src/coreclr/jit/lsraarm64.cpp:1834]:
```suggestion
//  `treeNode` that produces `registerCount` consecutive registers.
```

File: src/coreclr/jit/lsraarm64.cpp [hunk 3]
```diff
@@ -1816,6 +1829,57 @@ int LinearScan::BuildConsecutiveRegistersForUse(GenTree* treeNode, GenTree* rmwN
...
+//
+//  For the first RefPosition of the series, it sets the `regCount` field equal to
+//  the number of subsequent RefPositions (including the first one) involved for this
```
Comment (BruceForstall) [inline src/coreclr/jit/lsraarm64.cpp:1837]:
```suggestion
//  the total number of RefPositions (including the first one) involved for this
```

**Review by BruceForstall — APPROVED** [review]

**Comment by kunalspathak** [comment]
> @vargaz - any idea why `MinNumberPairwise` is failing in mono ?
> 
> ```
> 	  at <unknown> <0xffffffff>
> 	  at System.Object:__icall_wrapper_mono_gc_alloc_string <0x00007>
> 	  at System.Object:AllocString <0x00107>
> 	  at System.String:Concat <0x00103>
> 	  at TestLibrary.Generator:GetSingle <0x0006f>
> 	  at JIT.HardwareIntrinsics.Arm._AdvSimd.Arm64.SimpleBinaryOpTest__MinNumberPairwise_Vector64_Single:.ctor <0x002ff>
> 	  at JIT.HardwareIntrinsics.Arm._AdvSimd.Arm64.Program:MinNumberPairwise_Vector64_Single <0x0003f>
> 	  at Program:<<Main>$>g__TestExecutor360|0_361 <0x000f3>
> 	  at Program:<Main>$ <0x01973>
> 	  at <Module>:runtime_invoke_int_object <0x00087>
> ```

cc: @SamMonoRT @fanyang-mono - Is this the matter of adding an entry in `simd-methods.h`?

**Comment by vargaz** [comment]
> > @vargaz - any idea why `MinNumberPairwise` is failing in mono ?
> > ```
> > 	  at <unknown> <0xffffffff>
> > 	  at System.Object:__icall_wrapper_mono_gc_alloc_string <0x00007>
> > 	  at System.Object:AllocString <0x00107>
> > 	  at System.String:Concat <0x00103>
> > 	  at TestLibrary.Generator:GetSingle <0x0006f>
> > 	  at JIT.HardwareIntrinsics.Arm._AdvSimd.Arm64.SimpleBinaryOpTest__MinNumberPairwise_Vector64_Single:.ctor <0x002ff>
> > 	  at JIT.HardwareIntrinsics.Arm._AdvSimd.Arm64.Program:MinNumberPairwise_Vector64_Single <0x0003f>
> > 	  at Program:<<Main>$>g__TestExecutor360|0_361 <0x000f3>
> > 	  at Program:<Main>$ <0x01973>
> > 	  at <Module>:runtime_invoke_int_object <0x00087>
> > ```
> 
> cc: @SamMonoRT @fanyang-mono - Is this the matter of adding an entry in `simd-methods.h`?

I can reproduce. The crash appears unrelated to that specific test, the addition of the new LoadVectorX tests triggers it.

**Comment by vargaz** [comment]
Also, according to the test output:
Supported ISAs:
  AdvSimd:   False

The new tests are not even supposed to run.

**Comment by kunalspathak** [comment]
> Also, according to the test output: Supported ISAs: AdvSimd: False

Wondering why `AdvSimd: False` on linux-arm64 Release leg?

> The new tests are not even supposed to run.

That's right, it crashed while trying to run `RunUnsupportedScenario()`. 

> I can reproduce. The crash appears unrelated to that specific test, the addition of the new LoadVectorX tests triggers it.

I am assuming this PR cannot be merged until we fix this issue. Are you planning to submit a fix in this PR?

**Comment by vargaz** [comment]
> > Also, according to the test output: Supported ISAs: AdvSimd: False
> 
> Wondering why `AdvSimd: False` on linux-arm64 Release leg?
> 

AdvSimd is only supported on mono when running with llvm, not with the mono JIT.

> > The new tests are not even supposed to run.
> 
> That's right, it crashed while trying to run `RunUnsupportedScenario()`.
> 
> > I can reproduce. The crash appears unrelated to that specific test, the addition of the new LoadVectorX tests triggers it.
> 
> I am assuming this PR cannot be merged until we fix this issue. Are you planning to submit a fix in this PR?

Tracking this down will take some time. Would be nice to be able to disable this on mono, but I don't see any such functionality in GenerateHWIntrinsicTests_Arm.cs.

**Comment by vargaz** [comment]
The problem seems to be in the DataTable class in the tests:
```
            public DataTable({Op1BaseType}[] outArray1, {Op1BaseType}[] outArray2, {Op1BaseType}[] inArray, int alignment)
            {
                int sizeOfInArray = inArray.Length * Unsafe.SizeOf<{Op1BaseType}>();
                int sizeOfOutArray1 = outArray1.Length * Unsafe.SizeOf<{Op1BaseType}>();
                int sizeOfOutArray2 = outArray2.Length * Unsafe.SizeOf<{Op1BaseType}>();
                if ((alignment != 16 && alignment != 8) || (alignment * 2) < sizeOfInArray || (alignment * 2) < sizeOfOutArray1 || (alignment * 2) < sizeOfOutArray2)
                {
                    throw new ArgumentException("Invalid value of alignment");
                }

                this.inArray = new byte[alignment * 2];
                this.outArray1 = new byte[alignment * 2];
                this.outArray2 = new byte[alignment * 2];

                this.inHandle = GCHandle.Alloc(this.inArray, GCHandleType.Pinned);
                this.outHandle1 = GCHandle.Alloc(this.outArray1, GCHandleType.Pinned);
                this.outHandle2 = GCHandle.Alloc(this.outArray2, GCHandleType.Pinned);

                this.alignment = (ulong)alignment;

                Unsafe.CopyBlockUnaligned(ref Unsafe.AsRef<byte>(inArrayPtr), ref Unsafe.As<{Op1BaseType}, byte>(ref inArray[0]), (uint)sizeOfInArray);
            }

            public void* inArrayPtr => Align((byte*)(inHandle.AddrOfPinnedObject().ToPointer()), alignment);
```

If inArray is not 16 byte aligned, then the copy done by Unsafe.CopyBlockUnaligned () can write out of the bounds of the array, causing random crashes. This affects all the LoadVectorx<N>Test.template files. A fix would be to bump the
multiplier (i.e. 2 -> 3, 3 -> 4, 4 -> 5) in
``` this.inArray = new byte[alignment * 2];```

**Comment by kunalspathak** [comment]
>A fix would be to bump the multiplier (i.e. 2 -> 3, 3 -> 4, 4 -> 5) in

I am not sure why this is not problem for other test cases.

**Comment by kunalspathak** [comment]
Failures seems to be known: https://github.com/dotnet/runtime/pull/92855/checks?check_run_id=17415219399

**Review by kunalspathak — COMMENTED** [review]
LGTM. Just a request to update the documentation for LoadVector equivalent.

**Comment by kunalspathak** [comment]
> TM. Just a request to update the documentation for LoadVector equivalent.

Ah, wrong PR :)

**Pull request merged**